core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala

```diff
@@ -811,6 +811,87 @@ class AdminClientIntegrationTest extends IntegrationTestHarness with Logging {
       assertEquals(3, servers(i).replicaManager.getReplica(topicPartition).get.logStartOffset)
   }
 
+  @Test
+  def testReplicaCanFetchFromLogStartOffsetAfterDeleteRecords(): Unit = {
+    val leaders = createTopic(topic, numPartitions = 1, replicationFactor = serverCount)
+    val followerIndex = if (leaders(0) != servers(0).config.brokerId) 0 else 1
+
+    def waitForFollowerLSOandLEOAfterRestart(expectedLSO: Long, expectedLEO: Long): Unit = {
+      TestUtils.waitUntilTrue(() => servers(followerIndex).replicaManager.getReplica(topicPartition) != None,
+                              "Expected follower to create replica for partition")
+
+      // wait until the follower discovers that log start offset moved beyond its HW
+      TestUtils.waitUntilTrue(() => {
+        servers(followerIndex).replicaManager.getReplica(topicPartition).get.logStartOffset == expectedLSO
+      }, s"Expected follower to discover new log start offset $expectedLSO")
+
+      TestUtils.waitUntilTrue(() => {
+        servers(followerIndex).replicaManager.getReplica(topicPartition).get.logEndOffset.messageOffset == expectedLEO
+      }, s"Expected new replica to catch up to log end offset $expectedLEO")
+    }
+
+    // we will produce to topic and delete records while one follower is down
+    killBroker(followerIndex)
+
+    client = AdminClient.create(createConfig)
+    sendRecords(producers.head, 100, topicPartition)
+
+    val result = client.deleteRecords(Map(topicPartition -> RecordsToDelete.beforeOffset(3L)).asJava)
+    result.all().get()
+
+    // start the stopped broker to verify that it will be able to fetch from new log start offset
+    restartDeadBrokers()
+
+    waitForFollowerLSOandLEOAfterRestart(expectedLSO=3L, expectedLEO=100L)
+
+    // after the new replica caught up, all replicas should have same log start offset
+    for (i <- 0 until serverCount)
+      assertEquals(3, servers(i).replicaManager.getReplica(topicPartition).get.logStartOffset)
+
+    // kill the same follower again, produce more records, and delete records beyond follower's LOE
+    killBroker(followerIndex)
+    sendRecords(producers.head, 100, topicPartition)
+    val result1 = client.deleteRecords(Map(topicPartition -> RecordsToDelete.beforeOffset(117L)).asJava)
+    result1.all().get()
+    restartDeadBrokers()
+    waitForFollowerLSOandLEOAfterRestart(expectedLSO=117L, expectedLEO=200L)
+
+    client.close()
```

Comment:
nit: move to finally block so that a failure here won't cause other tests which are validating active threads to fail?

Reply:
Just checked... I was wrong to call `client.close()` here, because client.close() is called on test's `tearDown` which is called after every test even if a test fails.